src/peer-store/index.js

```diff
@@ -0,0 +1,190 @@
+'use strict'
+
+const assert = require('assert')
+const debug = require('debug')
+const log = debug('libp2p:peer-store')
+log.error = debug('libp2p:peer-store:error')
+
+const { EventEmitter } = require('events')
+
+const PeerInfo = require('peer-info')
+
+/**
+ * Responsible for managing known peers, as well as their addresses and metadata
+ * @fires PeerStore#peer Emitted when a peer is connected to this node
+ * @fires PeerStore#change:protocols
+ * @fires PeerStore#change:multiaddrs
+ */
+class PeerStore extends EventEmitter {
+  constructor () {
+    super()
+
+    /**
+     * Map of peers
+     *
+     * @type {Map<string, PeerInfo>}
+     */
+    this.peers = new Map()
+
+    // TODO: Track ourselves. We should split `peerInfo` up into its pieces so we get better
+    // control and observability. This will be the initial step for removing PeerInfo
+    // https://github.com/libp2p/go-libp2p-core/blob/master/peerstore/peerstore.go
+    // this.addressBook = new Map()
+    // this.protoBook = new Map()
+  }
+
+  /**
+   * Stores the peerInfo of a new peer.
+   * If already exist, its info is updated.
+   * @param {PeerInfo} peerInfo
+   */
+  put (peerInfo) {
+    assert(PeerInfo.isPeerInfo(peerInfo), 'peerInfo must be an instance of peer-info')
+
+    // Already know the peer?
+    if (this.peers.has(peerInfo.id.toB58String())) {
+      this.update(peerInfo)
+    } else {
+      this.add(peerInfo)
+
+      // Emit the new peer found
+      this.emit('peer', peerInfo)
+    }
+  }
+
+  /**
+   * Add a new peer to the store.
+   * @param {PeerInfo} peerInfo
+   */
+  add (peerInfo) {
+    assert(PeerInfo.isPeerInfo(peerInfo), 'peerInfo must be an instance of peer-info')
+
+    // Create new instance and add values to it
+    const newPeerInfo = new PeerInfo(peerInfo.id)
+
+    peerInfo.multiaddrs.forEach((ma) => newPeerInfo.multiaddrs.add(ma))
+    peerInfo.protocols.forEach((p) => newPeerInfo.protocols.add(p))
+
+    const connectedMa = peerInfo.isConnected()
+    connectedMa && newPeerInfo.connect(connectedMa)
+
+    const peerProxy = new Proxy(newPeerInfo, {
+      set: (obj, prop, value) => {
+        if (prop === 'multiaddrs') {
+          this.emit('change:multiaddrs', {
+            peerInfo: obj,
+            multiaddrs: value.toArray()
+          })
+        } else if (prop === 'protocols') {
+          this.emit('change:protocols', {
+            peerInfo: obj,
+            protocols: Array.from(value)
+          })
+        }
+        return true
```

Comment:
I believe you will need to call Reflect.set here for the changes to occur.
```suggestion
        return Reflect.set(obj, prop, value)
```